src/main/java/com/iota/iri/service/snapshot/impl/LocalSnapshotManagerImpl.java

```diff
@@ -0,0 +1,128 @@
+     */
+    private final SnapshotConfig config;
+
+    /**
+     * Holds a reference to the {@link ThreadIdentifier} for the monitor thread.
+     *
+     * Using a {@link ThreadIdentifier} for spawning the thread allows the {@link ThreadUtils} to spawn exactly one
+     * thread for this instance even when we call the {@link #start(MilestoneTracker)} method multiple times.
+     */
+    private ThreadIdentifier monitorThreadIdentifier = new ThreadIdentifier("Local Snapshots Monitor");
+
+    /**
+     * Creates the {@link LocalSnapshotManager} that takes care of automatically creating local snapshots when the
+     * defined intervals have passed.
+     *
+     * It simply stores the passed in parameters in their private properties.
+     *
+     * @param snapshotProvider data provider for the relevant {@link Snapshot} instances
+     * @param transactionPruner manager for the pruning jobs that allows us to clean up old transactions
+     * @param tangle object which acts as a database interface
+     * @param config configuration with important snapshot related parameters
+     */
+    public LocalSnapshotManagerImpl(SnapshotProvider snapshotProvider, TransactionPruner transactionPruner,
+            Tangle tangle, SnapshotConfig config) {
+
+        this.snapshotProvider = snapshotProvider;
+        this.transactionPruner = transactionPruner;
+        this.tangle = tangle;
+        this.config = config;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void start(MilestoneTracker milestoneTracker) {
+        ThreadUtils.spawnThread(() -> monitorThread(milestoneTracker), monitorThreadIdentifier);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void shutdown() {
+        ThreadUtils.stopThread(monitorThreadIdentifier);
+    }
+
+    /**
+     * This method contains the logic for the monitoring Thread.
+     *
+     * It periodically checks if a new {@link Snapshot} has to be taken until the {@link Thread} is terminated. If it
+     * detects that a {@link Snapshot} is due it triggers the creation of the {@link Snapshot} by calling
+     * {@link LocalSnapshotService#takeLocalSnapshot(Tangle, SnapshotProvider, SnapshotConfig, TransactionPruner)}.
+     *
+     * @param milestoneTracker tracker for the milestones to determine when a new local snapshot is due
+     */
+    private void monitorThread(MilestoneTracker milestoneTracker) {
+        while (!Thread.currentThread().isInterrupted()) {
+            int localSnapshotInterval = milestoneTracker.getStatus() == INITIALIZED &&
+                    snapshotProvider.getLatestSnapshot().getIndex() == milestoneTracker.latestMilestoneIndex
+                    ? config.getLocalSnapshotsIntervalSynced()
+                    : config.getLocalSnapshotsIntervalUnsynced();
+
+            if (snapshotProvider.getLatestSnapshot().getIndex() - snapshotProvider.getInitialSnapshot().getIndex() >
```

Comment:
getInitialSnapshot() and getLatestSnapshot() should never ever return null because it is a  very fundamental necessity for the node to have access to these instances.

The SnapshotProvider throws an exception if it can not load the instances and should cause IRI to stop.